### PR TITLE
[fix][txn] fix the consumer stuck due to deduplicated messages in pending ack state

### PR DESCRIPTION
**Pull request by hrzzzz**
### Motivation

https://github.com/apache/pulsar/pull/19581 filter messages in dispatcher. However, it filters out duplicate messages after calculating `totalMessages`, which results in occupying permits even though the messages were not sent to the consumer. This could potentially cause the consumer to be stuck and unable to consume.

### Modifications

Calculate totalMessages after filtering out the messages.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

- Added test in `org.apache.pulsar.client.impl.TransactionEndToEndTest#testMsgsInPendingAckStateWouldNotGetTheConsumerStuck`

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/hrzzzz/pulsar/pull/2

### CODE REVIEW

**Comment by hrzzzz** [comment]
@congbobo184 @aloyszhang PTAL

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TransactionEndToEndTest.java [hunk 1]
```diff
@@ -254,6 +254,62 @@ private void testFilterMsgsInPendingAckStateWhenConsumerDisconnect(boolean enabl
...
+                .receiverQueueSize(numStep3Receive)
+                .subscriptionName(subscription)
+                .subscriptionType(SubscriptionType.Shared)
+                .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
+                .subscribe();
+
+        // Step 3, try to consume the rest messages and should receive all of them
+        for (int i = 0; i < numStep3Receive; i++) {
+            // should get the message instead of timeout
+            Message<Integer> msg = consumer2.receive(10, TimeUnit.SECONDS);
```
Comment (congbobo184) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TransactionEndToEndTest.java:308]: only 3 is enough

**Review by congbobo184 — APPROVED** [review]
Good catch!

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TransactionEndToEndTest.java (hunk 1 above)
Comment (hrzzzz) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TransactionEndToEndTest.java]: fixed

**Review by liangyepianzhou — APPROVED** [review]

**Review by poorbarcode — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TransactionEndToEndTest.java [hunk 2]
```diff
@@ -254,6 +254,62 @@ private void testFilterMsgsInPendingAckStateWhenConsumerDisconnect(boolean enabl
         Assert.assertEquals(receiveCounter, count / 2);
     }
 
+    @Test
+    private void testMsgsInPendingAckStateWouldNotGetTheConsumerStuck() throws Exception {
+        final String topicName = NAMESPACE1 + "/testMsgsInPendingAckStateWouldNotGetTheConsumerStuck";
+        final String subscription = "test";
+
+        Producer<Integer> producer = pulsarClient.newProducer(Schema.INT32)
```
Comment (Technoboy-) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TransactionEndToEndTest.java:262]: Please add `@Cleanup` for `producer` and `consumer`

Comment (hrzzzz) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TransactionEndToEndTest.java:263]: fixed

**Pull request merged**